File: apps/desktop/src/shell/renderer/bridge/runtime-bridge/nimi-data-migration.ts
/**
 * Renderer bridge for the `nimi_data` directory-ownership + migration flow.
 *
 * Backend authority: `apps/desktop/src-tauri/src/nimi_data_migration/**`,
 * spec `P-MIG-006/007/008` in `local-config-migration-contract.md`.
 *
 * This is the typed renderer surface the later T10.3 Settings / T10.4 Support
 * UI waves call to drive the `P-MIG-007` data-root migration and the
 * `P-MIG-008` destructive-cleanup flow. It does NOT itself render any panel.
 */

import { hasTauriInvoke } from './env';
import { invokeChecked } from './invoke';

/** Terminal / contract state of a `nimi_data` migration (`P-MIG-007`). */
export type NimiDataMigrationState =
  | 'preview'
  | 'confirmed'
  | 'in_progress'
  | 'verifying'
  | 'completed'
  | 'failed'
  | 'repair_required';

/** One per-directory impact line of a migration preview. */
export interface NimiDataDirectoryImpact {
  readonly directory: string;
  readonly owner: string;
  readonly runtimeOwned: boolean;
  readonly declared: boolean;
  readonly totalBytes: number;
  readonly fileCount: number;
}

/** The `P-MIG-007` size / impact preview of a data-root migration. */
export interface NimiDataMigrationPreview {
  readonly sourceRoot: string;
  readonly targetRoot: string;
  readonly totalBytes: number;
  readonly totalFiles: number;
  readonly totalDirectories: number;
  readonly directories: readonly NimiDataDirectoryImpact[];
  readonly unownedDirectories: readonly string[];
  readonly includesRuntimeOwnedData: boolean;
}

/** The typed outcome of running a `nimi_data` data-root migration. */
export interface NimiDataMigrationOutcome {
  readonly state: NimiDataMigrationState;
  readonly previousRoot: string;
  readonly newRoot: string;
  readonly preview: NimiDataMigrationPreview;
  readonly verifiedBytes: number;
  readonly verifiedFiles: number;
  readonly verifiedDigest: string | null;
  readonly oldRootRetained: boolean;
  readonly error: string | null;
}

/** The `P-MIG-008` cleanup impact plan for a `nimi_data` directory. */
export interface NimiDataCleanupPlan {
  readonly directory: string;
  readonly owner: string;
  readonly cleanupClass: string;
  readonly totalBytes: number;
  readonly fileCount: number;
  readonly requiresConfirmation: boolean;
  readonly runtimeOwnerBlocked: boolean;
}

/** The result of a confirmed `nimi_data` cleanup. */
export interface NimiDataCleanupOutcome {
  readonly directory: string;
  readonly removedBytes: number;
  readonly removedFiles: number;
}

/**
 * The `P-MIG-008` impact preview for reclaiming a retained old `nimi_data`
 * data root. Reclaiming the old root is always destructive, so
 * `requiresConfirmation` is always `true`. `sameAsActive` / `activeNestedInOld`
 * are two fail-close conditions that make a reclaim impossible, and `recorded`
 * reflects whether the backend migration ledger actually retained this path;
 * `reclaimable` is `recorded && !sameAsActive && !activeNestedInOld`. When not
 * reclaimable the backend does not scan the path, so `totalBytes` / `fileCount`
 * are reported as `0` rather than probing an unauthorized location.
 */
export interface NimiDataOldRootReclaimPlan {
  readonly oldRoot: string;
  readonly activeRoot: string;
  readonly totalBytes: number;
  readonly fileCount: number;
  readonly requiresConfirmation: boolean;
  readonly sameAsActive: boolean;
  readonly activeNestedInOld: boolean;
  readonly recorded: boolean;
  readonly reclaimable: boolean;
}

/**
 * The explicit `P-MIG-008` destructive-cleanup confirmation token a UI must
 * collect from the user before a non-pure-cache cleanup. It mirrors the
 * backend `DESTRUCTIVE_CLEANUP_CONFIRMATION` constant.
 */
export const NIMI_DATA_DESTRUCTIVE_CLEANUP_CONFIRMATION = 'CLEAN';

function asRecord(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${label} returned invalid payload`);
  }
  return value as Record<string, unknown>;
}

const MIGRATION_STATES = new Set<NimiDataMigrationState>([
  'preview',
  'confirmed',
  'in_progress',
  'verifying',
  'completed',
  'failed',
  'repair_required',
]);

function parseDirectoryImpact(value: unknown): NimiDataDirectoryImpact {
  const record = asRecord(value, 'nimi_data directory impact');
  return {
    directory: String(record.directory || ''),
    owner: String(record.owner || ''),
    runtimeOwned: record.runtimeOwned === true,
    declared: record.declared === true,
    totalBytes: Number(record.totalBytes || 0),
    fileCount: Number(record.fileCount || 0),
  };
}

function parsePreview(value: unknown): NimiDataMigrationPreview {
  const record = asRecord(value, 'nimi_data migration preview');
  return {
    sourceRoot: String(record.sourceRoot || ''),
    targetRoot: String(record.targetRoot || ''),
    totalBytes: Number(record.totalBytes || 0),
    totalFiles: Number(record.totalFiles || 0),
    totalDirectories: Number(record.totalDirectories || 0),
    directories: Array.isArray(record.directories)
      ? record.directories.map(parseDirectoryImpact)
      : [],
    unownedDirectories: Array.isArray(record.unownedDirectories)
      ? record.unownedDirectories.map((entry) => String(entry || '')).filter(Boolean)
      : [],
    includesRuntimeOwnedData: record.includesRuntimeOwnedData === true,
  };
}

function parseOutcome(value: unknown): NimiDataMigrationOutcome {
  const record = asRecord(value, 'nimi_data migration outcome');
  const state = String(record.state || '') as NimiDataMigrationState;
  if (!MIGRATION_STATES.has(state)) {
    throw new Error(`nimi_data migration returned invalid state: ${state}`);
  }
  const verifiedDigest = record.verifiedDigest;
  return {
    state,
    previousRoot: String(record.previousRoot || ''),
    newRoot: String(record.newRoot || ''),
    preview: parsePreview(record.preview),
    verifiedBytes: Number(record.verifiedBytes || 0),
    verifiedFiles: Number(record.verifiedFiles || 0),
    verifiedDigest:
      typeof verifiedDigest === 'string' && verifiedDigest.trim() ? verifiedDigest : null,
    oldRootRetained: record.oldRootRetained === true,
    error: typeof record.error === 'string' && record.error.trim() ? record.error : null,
  };
}

function parseCleanupPlan(value: unknown): NimiDataCleanupPlan {
  const record = asRecord(value, 'nimi_data cleanup plan');
  return {
    directory: String(record.directory || ''),
    owner: String(record.owner || ''),
    cleanupClass: String(record.cleanupClass || ''),
    totalBytes: Number(record.totalBytes || 0),
    fileCount: Number(record.fileCount || 0),
    requiresConfirmation: record.requiresConfirmation === true,
    runtimeOwnerBlocked: record.runtimeOwnerBlocked === true,
  };
}

function parseCleanupOutcome(value: unknown): NimiDataCleanupOutcome {
  const record = asRecord(value, 'nimi_data cleanup outcome');
  return {
    directory: String(record.directory || ''),
    removedBytes: Number(record.removedBytes || 0),
    removedFiles: Number(record.removedFiles || 0),
  };
}

function parseOldRootReclaimPlan(value: unknown): NimiDataOldRootReclaimPlan {
  const record = asRecord(value, 'nimi_data old-root reclaim plan');
  return {
    oldRoot: String(record.oldRoot || ''),
    activeRoot: String(record.activeRoot || ''),
    totalBytes: Number(record.totalBytes || 0),
    fileCount: Number(record.fileCount || 0),
    requiresConfirmation: record.requiresConfirmation === true,
    sameAsActive: record.sameAsActive === true,
    activeNestedInOld: record.activeNestedInOld === true,
    recorded: record.recorded === true,
    reclaimable: record.reclaimable === true,
  };
}

/**
 * `P-MIG-007` preview: compute the size / impact preview for moving the
 * current `nimi_data` data root to `targetRoot`. Moves nothing.
 */
export async function previewNimiDataMigration(
  targetRoot: string,
): Promise<NimiDataMigrationPreview> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_migration_preview requires Tauri runtime');
  }
  return invokeChecked('nimi_data_migration_preview', { payload: { targetRoot } }, parsePreview);
}

/**
 * `P-MIG-007` run: execute a confirmed `nimi_data` data-root migration —
 * staged integrity-checked copy, atomic promote, pointer cutover last.
 *
 * The caller MUST have shown the user the preview and obtained an explicit
 * confirmation before invoking this. On a `completed` outcome the caller must
 * follow up with the Runtime `config.json` `dataRootRef` re-sync.
 */
export async function runNimiDataMigration(
  targetRoot: string,
): Promise<NimiDataMigrationOutcome> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_migration_run requires Tauri runtime');
  }
  return invokeChecked('nimi_data_migration_run', { payload: { targetRoot } }, parseOutcome);
}

/** `P-MIG-008` plan: compute the cleanup impact of a `nimi_data` directory. */
export async function planNimiDataCleanup(directory: string): Promise<NimiDataCleanupPlan> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_cleanup_plan requires Tauri runtime');
  }
  return invokeChecked('nimi_data_cleanup_plan', { directory }, parseCleanupPlan);
}

/**
 * `P-MIG-008` execute: run a confirmed cleanup of a `nimi_data` directory.
 * `confirmation` must equal {@link NIMI_DATA_DESTRUCTIVE_CLEANUP_CONFIRMATION}
 * for any non-pure-cache directory.
 */
export async function executeNimiDataCleanup(
  directory: string,
  confirmation?: string,
): Promise<NimiDataCleanupOutcome> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_cleanup_execute requires Tauri runtime');
  }
  return invokeChecked(
    'nimi_data_cleanup_execute',
    { payload: { directory, confirmation: confirmation ?? null } },
    parseCleanupOutcome,
  );
}

/**
 * `P-MIG-008` plan: compute the impact preview for reclaiming a retained old
 * `nimi_data` data root (`MigrationOutcome.previousRoot`, when
 * `oldRootRetained` is `true`). Deletes nothing. The active (protected) data
 * root is resolved backend-side, never supplied by the renderer.
 */
export async function planNimiDataOldRootReclaim(
  oldRoot: string,
): Promise<NimiDataOldRootReclaimPlan> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_old_root_reclaim_plan requires Tauri runtime');
  }
  return invokeChecked(
    'nimi_data_old_root_reclaim_plan',
    { payload: { oldRoot } },
    parseOldRootReclaimPlan,
  );
}

/**
 * `P-MIG-008` execute: reclaim a retained old `nimi_data` data root. Always
 * destructive, so `confirmation` must equal
 * {@link NIMI_DATA_DESTRUCTIVE_CLEANUP_CONFIRMATION}. Fails closed if the old
 * root resolves to the active data root or the active root is nested inside it.
 */
export async function reclaimNimiDataOldRoot(
  oldRoot: string,
  confirmation?: string,
): Promise<NimiDataCleanupOutcome> {
  if (!hasTauriInvoke()) {
    throw new Error('nimi_data_old_root_reclaim_execute requires Tauri runtime');
  }
  return invokeChecked(
    'nimi_data_old_root_reclaim_execute',
    { payload: { oldRoot, confirmation: confirmation ?? null } },
    parseCleanupOutcome,
  );
}

File: apps/desktop/src-tauri/src/nimi_data_migration.rs
//! `nimi_data` directory ownership enforcement + data-root migration flow.
//!
//! Spec authority: `.nimi/spec/platform/kernel/local-config-migration-contract.md`
//! `P-MIG-006`, `P-MIG-007`, `P-MIG-008`, and the kernel table
//! `.nimi/spec/platform/kernel/tables/nimi-data-directory-ownership.yaml`.
//!
//! This is the T10.2 wave: the `nimi_data` directory / migration mechanism.
//!
//! - `P-MIG-006` (`ownership` + `layout`): the 15-row `nimi_data` directory
//!   owner + cleanup matrix as the authoritative directory model — directory
//!   creation routes through `enforce_data_root_layout`, and every cleanup /
//!   ownership check resolves against the matrix.
//! - `P-MIG-007` (`preview` + `copy` + `flow`): a real data-root migration
//!   flow — size / impact preview, integrity-checked copy, a typed migration
//!   state machine, atomic pointer cutover committed last, and no orphaning of
//!   the old or new location on any failure.
//! - `P-MIG-008` (`cleanup`): any destructive cleanup requires an explicit
//!   confirmation token and obeys the matrix's per-directory cleanup rule;
//!   without the token a non-pure-cache cleanup fails closed.
//!
//! The flow + cleanup mechanics are exposed as a minimal Tauri command surface
//! the later T10.3 Settings / T10.4 Support UI waves call. This wave does NOT
//! build those panels.
//!
//! It reuses the T10.1 `local_config_migration` idioms (typed outcomes,
//! fail-closed, backup-before-mutate, atomic write) where they fit; it does
//! NOT redefine the `~/.nimi` config framework or the Runtime `config.json`
//! framework.

mod cleanup;
mod copy;
mod flow;
mod layout;
mod ownership;
mod preview;

#[cfg(test)]
mod tests;

pub use layout::enforce_data_root_layout;

use serde::Deserialize;

use cleanup::{
    execute_directory_cleanup, plan_directory_cleanup, plan_old_root_reclaim, reclaim_old_root,
    CleanupOutcome, CleanupPlan, OldRootReclaimPlan,
};
use flow::{preview_migration, run_migration, MigrationOutcome};
use preview::MigrationPreview;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NimiDataMigrationPreviewPayload {
    /// Absolute target `nimi_data` path the user wants to move the data root
    /// to.
    pub target_root: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NimiDataMigrationRunPayload {
    /// Absolute target `nimi_data` path. The renderer obtains this from a
    /// preview the user has explicitly confirmed.
    pub target_root: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NimiDataCleanupPayload {
    /// First-level `nimi_data` directory id from the `P-MIG-006` matrix.
    pub directory: String,
    /// Explicit `P-MIG-008` confirmation token. Required for any
    /// non-pure-cache directory; ignored for a pure-cache directory.
    pub confirmation: Option<String>,
}

/// `P-MIG-007` preview: compute the size / impact preview for moving the
/// current `nimi_data` data root to `target_root`. Moves nothing.
#[tauri::command]
pub async fn nimi_data_migration_preview(
    payload: NimiDataMigrationPreviewPayload,
) -> Result<MigrationPreview, String> {
    run_blocking(move || {
        let source = crate::desktop_product_control::selected_product_data_root()?;
        preview_migration(&source, &payload.target_root)
    })
    .await
}

/// `P-MIG-007` run: execute a confirmed `nimi_data` data-root migration end to
/// end — staged integrity-checked copy, atomic promote, and pointer cutover
/// committed last. Returns the typed `MigrationOutcome` state machine result.
///
/// On a `completed` outcome the renderer must follow up with the Runtime
/// `config.json` `dataRootRef` re-sync (a `K-CFG-*`-owned mechanism); this
/// command intentionally does not perform that — `P-MIG-007` keeps the
/// pointer-commit boundary at `~/.nimi/nimi.json`.
#[tauri::command]
pub async fn nimi_data_migration_run(
    payload: NimiDataMigrationRunPayload,
) -> Result<MigrationOutcome, String> {
    run_blocking(move || {
        let source = crate::desktop_product_control::selected_product_data_root()?;
        run_migration(&source, &payload.target_root)
    })
    .await
}

/// `P-MIG-008` plan: compute the cleanup impact preview for a first-level
/// `nimi_data` directory. Deletes nothing.
#[tauri::command]
pub async fn nimi_data_cleanup_plan(directory: String) -> Result<CleanupPlan, String> {
    run_blocking(move || {
        let data_root = crate::desktop_product_control::selected_product_data_root()?;
        plan_directory_cleanup(&data_root, &directory)
    })
    .await
}

/// `P-MIG-008` execute: run a confirmed cleanup of a first-level `nimi_data`
/// directory. Fails closed without the confirmation token for a non-pure-cache
/// directory; rejects Runtime-owned directories outright.
#[tauri::command]
pub async fn nimi_data_cleanup_execute(
    payload: NimiDataCleanupPayload,
) -> Result<CleanupOutcome, String> {
    run_blocking(move || {
        let data_root = crate::desktop_product_control::selected_product_data_root()?;
        execute_directory_cleanup(
            &data_root,
            &payload.directory,
            payload.confirmation.as_deref(),
        )
    })
    .await
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NimiDataOldRootReclaimPlanPayload {
    /// Absolute path of the retained old `nimi_data` data root to preview a
    /// reclaim for. The renderer obtains this from a completed `MigrationOutcome`
    /// (`previousRoot`, with `oldRootRetained = true`).
    pub old_root: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NimiDataOldRootReclaimPayload {
    /// Absolute path of the retained old `nimi_data` data root to reclaim.
    pub old_root: String,
    /// Explicit `P-MIG-008` confirmation token. Required — reclaiming the old
    /// root is always destructive.
    pub confirmation: Option<String>,
}

/// `P-MIG-008` plan: compute the impact preview for reclaiming a retained old
/// `nimi_data` data root. Deletes nothing.
///
/// Two backend-authoritative facts gate the reclaim and cannot be spoofed by
/// the renderer:
/// - the active (protected) data root is `selected_product_data_root()`, so the
///   guard against deleting the live data root cannot be bypassed;
/// - `recorded` comes from the migration reclaim ledger, so only a path a real
///   completed migration actually retained is ever marked reclaimable.
#[tauri::command]
pub async fn nimi_data_old_root_reclaim_plan(
    payload: NimiDataOldRootReclaimPlanPayload,
) -> Result<OldRootReclaimPlan, String> {
    run_blocking(move || {
        let old_path = std::path::PathBuf::from(payload.old_root.trim());
        let active_root = crate::desktop_product_control::selected_product_data_root()?;
        let recorded = crate::desktop_product_control::is_recorded_retained_old_root(&old_path)?;
        plan_old_root_reclaim(&old_path, &active_root, recorded)
    })
    .await
}

/// `P-MIG-008` execute: reclaim a retained old `nimi_data` data root. Always
/// requires the confirmation token; refuses to touch the active data root or a
/// root the active root is nested inside. Fails closed otherwise.
///
/// Authorization: the `old_root` MUST be a path the migration reclaim ledger
/// recorded (`is_recorded_retained_old_root`). A renderer-supplied path the
/// backend never retained can never drive a destructive delete — the `CLEAN`
/// token is a deliberate-action gate, not an authorization capability. The
/// active (protected) data root is the backend-authoritative
/// `selected_product_data_root()`, not a renderer-supplied value. On success
/// the ledger entry is consumed so the path is no longer advertised.
#[tauri::command]
pub async fn nimi_data_old_root_reclaim_execute(
    payload: NimiDataOldRootReclaimPayload,
) -> Result<CleanupOutcome, String> {
    run_blocking(move || {
        let old_path = std::path::PathBuf::from(payload.old_root.trim());
        let active_root = crate::desktop_product_control::selected_product_data_root()?;
        if !crate::desktop_product_control::is_recorded_retained_old_root(&old_path)? {
            return Err(format!(
                "拒绝回收：{} 不是已记录的迁移保留旧 nimi_data 数据根，无法回收",
                old_path.display()
            ));
        }
        let outcome = reclaim_old_root(&old_path, &active_root, payload.confirmation.as_deref())?;
        crate::desktop_product_control::consume_retained_old_root(&old_path)?;
        Ok(outcome)
    })
    .await
}

/// Run a blocking `nimi_data` filesystem operation off the async runtime.
///
/// The migration copy / scan / cleanup are blocking filesystem work; running
/// them on a Tauri async command thread would stall the event loop. This wraps
/// them in `spawn_blocking`, matching the other filesystem-heavy store tasks.
async fn run_blocking<T, F>(operation: F) -> Result<T, String>
where
    T: Send + 'static,
    F: FnOnce() -> Result<T, String> + Send + 'static,
{
    tauri::async_runtime::spawn_blocking(operation)
        .await
        .map_err(|error| format!("NIMI_DATA_MIGRATION_TASK_JOIN_FAILED: {error}"))?
}
